yarn-project/archiver/src/archiver/archiver.ts

```
@@ -878,7 +878,21 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
       const publishedCheckpoints = await Promise.all(retrievedCheckpoints.map(b => retrievedToPublishedCheckpoint(b)));
       const validCheckpoints: PublishedCheckpoint[] = [];
 
-      for (const published of publishedCheckpoints) {
+      for (let i = 0; i < retrievedCheckpoints.length; i++) {
+        const retrieved = retrievedCheckpoints[i];
+        const { published, error: publishedCheckpointError } = publishedCheckpoints[i];
+        if (!published) {
+          this.log.warn(
+            `Skipping checkpoint ${retrieved.checkpointNumber} due to inability to convert it to a published checkpoint`,
+            {
+              l1BlockNumber: retrieved.l1.blockNumber,
+              reason: publishedCheckpointError,
+            },
+          );
+
+          continue;
```

IIUC we used to throw if there was an error converting the checkpoint object, which halted the archiver sync, whereas we are now continuing, right? Why the change? Not saying we shouldn't, just wondering about the rationale for it since it's a big change. Also, do we know what could trigger an error in the conversion here (aside from a bug)?

---

I changed it this way to match what we do when the attestations are invalid. And when processing the logs from  L1 in `data_retrieval.ts`, we also simply "skip" the checkpoint if the archive roots do not match (although I'm not sure if that's possible).

If the proposer is honest, it's most likely a bug that triggers an error from the conversion. Same for the inHash below.

Let me know if you think throwing an error is preferable and I'll change it!

---

We purposefully skip the block if attestations are invalid since we know that the next proposer will "invalidate" this block, removing it from the chain. And we skip the block if the archive root does not match the one in rollup storage because it means we're trying to sync a block that has been pruned. In both situations the block we're skipping is not part of the chain.

That's why I'm worried about skipping in other situations. If we have a bug and there's a _valid_ block we can't sync, then we should not skip it, since the following block would fail due to the gap in the chain. So I think I'd go back to failing loudly. WDYT?

---

By the way, I'll approve so you can merge this one if you want to change the error handling in another PR. Whatever you prefer!

---

Makes perfect sense! I'll change the encoding and in hash to throw an error instead. Thanks for explaining it 🙌

---

Since you're at it, can you add a comment explaining the above (or just copy pasting it)? I realized that logic is not documented anywhere, my fault.